imap_processing/mag/l1a/mag_l1a_data.py

```diff
@@ -0,0 +1,259 @@
+        computed by adding 1/vectors_per_second to the previous vector's time.
+
+        This replaces self.vectors with a list of Vector objects.
+        """
+        sample_time_interval = 1 / self.vectors_per_second
+        previous_time = self.start_time
+        for index, vector in enumerate(self.vectors):
+            if index == 0:
+                new_vector = Vector(vector, self.start_time)
+            else:
+                new_vector = Vector(vector, previous_time + sample_time_interval)
+
+            previous_time = new_vector.timestamp
+
+            self.vectors[index] = new_vector
+
+    @staticmethod
+    def process_vector_data(
+        vector_data: np.ndarray, primary_count: int, secondary_count: int
+    ) -> (list[tuple], list[tuple]):
+        """
+        Given raw packet data, process into Vectors.
+
+        Vectors are grouped into primary sensor and secondary sensor, and returned as a
+        tuple (primary sensor vectors, secondary sensor vectors)
+
+        Written by MAG instrument team
+
+        Parameters
+        ----------
+        vector_data : np.ndarray
+            Raw vector data, in bytes. Contains both primary and secondary vector data
+            (first primary, then secondary)
+        primary_count : int
+            Count of the number of primary vectors
+        secondary_count : int
+            Count of the number of secondary vectors
+
+        Returns
+        -------
+        (primary, secondary)
+            Two arrays, each containing tuples of (x, y, z, sample_range) for each
+            vector sample.
+        """
+
+        # TODO: error handling
+        def to_signed16(n):
+            n = n & 0xFFFF
+            return n | (-(n & 0x8000))
+
+        pos = 0
+        primary_vectors = []
+        secondary_vectors = []
+
+        # Since the vectors are stored as 50 bit chunks but accessed via hex (4 bit
+        # chunks) there is some shifting required for processing the bytes.
+        # However, from a bit processing perspective, the first 48 bits of each 50 bit
+        # chunk corresponds to 3 16 bit signed integers. The last 2 bits are the sensor
+        # range.
+
+        for i in range(primary_count + secondary_count):  # 0..63 say
+            x, y, z, rng = 0, 0, 0, 0
+            if i % 4 == 0:  # start at bit 0, take 8 bits + 8bits
+                # pos = 0, 25, 50...
+                x = (
+                    ((vector_data[pos + 0] & 0xFF) << 8)
+                    | ((vector_data[pos + 1] & 0xFF) << 0)
+                ) & 0xFFFF
+                y = (
+                    ((vector_data[pos + 2] & 0xFF) << 8)
+                    | ((vector_data[pos + 3] & 0xFF) << 0)
+                ) & 0xFFFF
+                z = (
+                    ((vector_data[pos + 4] & 0xFF) << 8)
+                    | ((vector_data[pos + 5] & 0xFF) << 0)
+                ) & 0xFFFF
+                rng = (vector_data[pos + 6] >> 6) & 0x3
+                pos += 6
```

Comment:
Again minor (and I realize you got this from elsewhere), but I would probably write this with variables being parameterized and then used in the lookups a single time to try and avoid the duplication of those lookups. No need to change here unless you like this form better.
```python
if i % 4 == 0:
    left_mask = 0xFF
    left_shift = 0
elif i % 4 == 1:
    left_mask = 0x3F
    left_shift = 2
...
x = ((vector_data[pos + 0] & left_mask) << (left_shift + 8))
        | (vector_data[pos + 1] & 0xFF) << left_shift))
       & 0xFFFF
```